File: vod_sync.py
import os
import sys
import json
import re
import urllib.request
import tempfile
import shutil
import subprocess
import numpy as np
from scipy.io import wavfile
from scipy.signal import correlate
from twitchsync import TwitchSync

SAMPLE_RATE = 44100
CLIP_SECONDS = 10

def parse_twitch_url(url):
    video_id_match = re.search(r'videos/(\d+)', url)
    time_match     = re.search(r'[?&]t=([\dhms]+)', url)
    clip_match     = re.search(r'/clip/([\w-]+)', url)

    video_id   = video_id_match.group(1) if video_id_match else None
    clip_slug  = clip_match.group(1)     if clip_match     else None
    ts_str     = time_match.group(1)     if time_match     else "0s"

    h_match = re.search(r'(\d+)h', ts_str)
    m_match = re.search(r'(\d+)m', ts_str)
    s_match = re.search(r'(\d+)s', ts_str)

    h = int(h_match.group(1)) if h_match else 0
    m = int(m_match.group(1)) if m_match else 0
    s = int(s_match.group(1)) if s_match else 0
    
    total_seconds = h * 3600 + m * 60 + s
    return video_id, ts_str, total_seconds, clip_slug

def seconds_to_twitch_ts(total_seconds):
    total_seconds = max(0, round(total_seconds))
    h = total_seconds // 3600
    m = (total_seconds % 3600) // 60
    s = total_seconds % 60
    return f"{h}h{m}m{s}s"

def build_twitch_vod_url(video_id, ts_str):
    return f"https://www.twitch.tv/videos/{video_id}?t={ts_str}"

def _yt_dlp_available():
    return shutil.which("yt-dlp") is not None or shutil.which("yt_dlp") is not None

def download_audio_segment(vod_url, start_sec, out_path, clip_dur=CLIP_SECONDS):
    clip_start = max(0, start_sec - CLIP_SECONDS // 2)
    clip_end   = clip_start + CLIP_SECONDS
    section_arg = f"*{clip_start}-{clip_end}"

    cmd = [
        "yt-dlp",
        vod_url,
        "--no-playlist",
        "--quiet",
        "--no-warnings",
        "--format", "bestaudio/best",
        "--download-sections", section_arg,
        "--postprocessor-args", f"ffmpeg:-ar {SAMPLE_RATE} -ac 1",
        "--extract-audio",
        "--audio-format", "wav",
        "--output", out_path,
        "--force-overwrites",
    ]

    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        raise RuntimeError(f"yt-dlp failed for {vod_url}:\n{result.stderr}")

    if not os.path.exists(out_path):
        candidate = out_path + ".wav"
        if os.path.exists(candidate):
            os.rename(candidate, out_path)
        else:
            raise FileNotFoundError(f"yt-dlp didn't produce expected file at {out_path}")

def calculate_audio_offset(vod1_url, time1_sec, vod2_url, time2_sec):
    if not _yt_dlp_available():
        return 0.0

    tmp_dir = tempfile.mkdtemp(prefix="twitchsync_")
    wav1 = os.path.join(tmp_dir, "vod1.wav")
    wav2 = os.path.join(tmp_dir, "vod2.wav")

    try:
        download_audio_segment(vod1_url, time1_sec, wav1)
        download_audio_segment(vod2_url, time2_sec, wav2)

        rate1, data1 = wavfile.read(wav1)
        rate2, data2 = wavfile.read(wav2)

        data1 = data1.astype(np.float32) / np.iinfo(data1.dtype).max
        data2 = data2.astype(np.float32) / np.iinfo(data2.dtype).max

        if data1.ndim > 1: data1 = data1.mean(axis=1)
        if data2.ndim > 1: data2 = data2.mean(axis=1)

        min_len = min(len(data1), len(data2))
        data1 = data1[:min_len]
        data2 = data2[:min_len]

        correlation = correlate(data1, data2, mode='full')
        lags        = np.arange(-(min_len - 1), min_len)
        best_lag    = lags[np.argmax(correlation)]

        sr = rate1 if rate1 > 0 else SAMPLE_RATE
        return float(best_lag / sr)
    finally:
        shutil.rmtree(tmp_dir, ignore_errors=True)

def get_streamer_from_api(video_id, client_id, client_secret):
    try:
        auth_url = "https://id.twitch.tv/oauth2/token"
        body = f"client_id={client_id}&client_secret={client_secret}&grant_type=client_credentials".encode('utf-8')
        with urllib.request.urlopen(auth_url, data=body) as response:
            auth_data = json.loads(response.read().decode())
            token = auth_data['access_token']
        
        video_url = f"https://api.twitch.tv/helix/videos?id={video_id}"
        req = urllib.request.Request(video_url, headers={
            "Client-ID": client_id,
            "Authorization": f"Bearer {token}"
        })
        with urllib.request.urlopen(req) as response:
            video_data = json.loads(response.read().decode())
            if video_data['data'] and 'user_name' in video_data['data'][0]:
                return video_data['data'][0]['user_name']
    except:
        pass
    return None

def main():
    try:
        input_data = sys.stdin.read()
        if not input_data:
            return
        
        req = json.loads(input_data)
        url = req.get("url")
        streamers = req.get("streamers", [])
        
        client_id = os.environ.get("TWITCH_CLIENT_ID", "")
        client_secret = os.environ.get("TWITCH_CLIENT_SECRET", "")
        client = TwitchSync(client_id=client_id, client_secret=client_secret)
        matches = client.get_matches_for_all_streamers(streamers, url)
        
        # Determine source label
        source_label = "Source VOD"
        url_clean = url.replace('https://', '').replace('http://', '').replace('www.', '')
        ch_match = re.search(r'twitch\.tv/([^/?#]+)', url_clean)
        
        if ch_match:
            candidate = ch_match.group(1)
            if candidate != 'videos':
                source_label = candidate
            else:
                # If it's just /videos/ID, try to fetch the streamer name from the API
                video_id, _, _, _ = parse_twitch_url(url)
                if video_id and client_id and client_secret:
                    api_name = get_streamer_from_api(video_id, client_id, client_secret)
                    if api_name:
                        source_label = api_name

        matches.insert(0, {'streamer': source_label, 'result': url})

        for m in matches:
            _, ts_str, ts_sec, _ = parse_twitch_url(m['result'])
            m['refined_ts'] = ts_str
            m['ts_sec'] = ts_sec
            m['offset'] = 0.0
            m['total_offset'] = 0.0

        if len(matches) >= 2:
            vid_src, ts_str_src, ts_sec_src, _ = parse_twitch_url(matches[0]['result'])
            if vid_src:
                src_vod_url = matches[0]['result']
                for m in matches[1:]:
                    vid2, ts_str2, ts_sec2, clip2 = parse_twitch_url(m['result'])
                    if not vid2: continue
                    try:
                        offset = calculate_audio_offset(src_vod_url, ts_sec_src, m['result'], ts_sec2)
                        corrected_sec = ts_sec2 - offset
                        corrected_ts  = seconds_to_twitch_ts(corrected_sec)
                        m['refined_ts'] = corrected_ts
                        m['offset']     = offset
                        m['total_offset'] = corrected_sec - ts_sec_src
                        m['result'] = build_twitch_vod_url(vid2, corrected_ts)
                    except Exception:
                        pass
        
        # Format the output items nicely
        configs = []
        for match in matches:
            video_id, _, _, clip_slug = parse_twitch_url(match['result'])
            cfg = {"label": match['streamer']}
            if video_id:
                cfg["video"] = video_id
                cfg["time"]  = match.get('refined_ts')
                cfg["offset"] = match.get('offset', 0)
                cfg["total_offset"] = match.get('total_offset', 0)
            elif clip_slug:
                cfg["clip"] = clip_slug
            configs.append(cfg)
            
        print(json.dumps({"success": True, "data": configs}))
    except Exception as e:
        print(json.dumps({"success": False, "error": str(e)}))

if __name__ == "__main__":
    main()

File: internal/api/vod_sync.go
package api

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"math"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"
)

var (
	twitchToken      string
	twitchTokenMutex sync.Mutex
	twitchTokenExp   time.Time
)

type TwitchFollowedResponse struct {
	UserName        string `json:"user_name"`
	UserLogin       string `json:"user_login"`
	ProfileImageURL string `json:"profile_image_url"`
	IsLive          bool   `json:"is_live"`
	ViewerCount     int    `json:"viewer_count"`
	Title           string `json:"title"`
	GameName        string `json:"game_name"`
	VideoID         string `json:"video_id,omitempty"`
}

func GetTwitchToken(clientID, clientSecret string) (string, error) {
	twitchTokenMutex.Lock()
	defer twitchTokenMutex.Unlock()

	if twitchToken != "" && time.Now().Before(twitchTokenExp) {
		return twitchToken, nil
	}

	authURL := "https://id.twitch.tv/oauth2/token"
	data := url.Values{}
	data.Set("client_id", clientID)
	data.Set("client_secret", clientSecret)
	data.Set("grant_type", "client_credentials")

	req, err := http.NewRequest("POST", authURL, strings.NewReader(data.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("failed to get token: %s", string(body))
	}

	var res struct {
		AccessToken string `json:"access_token"`
		ExpiresIn   int    `json:"expires_in"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
		return "", err
	}

	twitchToken = res.AccessToken
	twitchTokenExp = time.Now().Add(time.Duration(res.ExpiresIn-60) * time.Second)
	return twitchToken, nil
}

type VodSyncRequest struct {
	Url       string   `json:"url"`
	Streamers []string `json:"streamers"`
}

type VodSyncConfig struct {
	Label       string  `json:"label"`
	Video       string  `json:"video,omitempty"`
	Clip        string  `json:"clip,omitempty"`
	Time        string  `json:"time,omitempty"`
	Offset      float64 `json:"offset"`
	TotalOffset float64 `json:"total_offset"`
}

type TwitchVideoResp struct {
	Data []struct {
		Id        string `json:"id"`
		UserId    string `json:"user_id"`
		UserName  string `json:"user_name"`
		CreatedAt string `json:"created_at"`
		Duration  string `json:"duration"`
	} `json:"data"`
}

type TwitchClipResp struct {
	Data []struct {
		Id              string `json:"id"`
		VideoId         string `json:"video_id"`
		VodOffset       int    `json:"vod_offset"`
		CreatedAt       string `json:"created_at"`
		BroadcasterName string `json:"broadcaster_name"`
	} `json:"data"`
}

func parseDurationStr(durationStr string) float64 {
	h, m, s := 0, 0, 0

	hIdx := strings.Index(durationStr, "h")
	if hIdx != -1 {
		h, _ = strconv.Atoi(durationStr[:hIdx])
		durationStr = durationStr[hIdx+1:]
	}

	mIdx := strings.Index(durationStr, "m")
	if mIdx != -1 {
		m, _ = strconv.Atoi(durationStr[:mIdx])
		durationStr = durationStr[mIdx+1:]
	}

	sIdx := strings.Index(durationStr, "s")
	if sIdx != -1 {
		s, _ = strconv.Atoi(durationStr[:sIdx])
	}

	return float64(h*3600 + m*60 + s)
}

func parseTwitchUrl(apiUrl string) (string, string, float64, string) {
	videoIdRe := regexp.MustCompile(`videos/(\d+)`)
	timeRe := regexp.MustCompile(`[?&]t=([\dhms]+)`)
	clipRe := regexp.MustCompile(`/clip/([^?&]+)`)
	clipsDomRe := regexp.MustCompile(`clips\.twitch\.tv/([^?&]+)`)

	var videoId, timeStr, clipSlug string
	var totalSeconds float64

	if m := videoIdRe.FindStringSubmatch(apiUrl); len(m) > 1 {
		videoId = m[1]
	}
	if m := timeRe.FindStringSubmatch(apiUrl); len(m) > 1 {
		timeStr = m[1]
		totalSeconds = parseDurationStr(timeStr)
	} else {
		timeStr = "0s"
	}

	if m := clipRe.FindStringSubmatch(apiUrl); len(m) > 1 {
		clipSlug = m[1]
	} else if m := clipsDomRe.FindStringSubmatch(apiUrl); len(m) > 1 {
		clipSlug = m[1]
	}

	return videoId, timeStr, totalSeconds, clipSlug
}

func formatDurationStr(totalSeconds float64) string {
	ts := int(math.Max(0, math.Round(totalSeconds)))
	h := ts / 3600
	m := (ts % 3600) / 60
	s := ts % 60
	return fmt.Sprintf("%dh%dm%ds", h, m, s)
}

func doTwitchGet(endpoint string, clientID, token string, result interface{}) error {
	req, err := http.NewRequest("GET", endpoint, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Client-ID", clientID)
	req.Header.Set("Authorization", "Bearer "+token)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("twitch api error (status %d): %s", resp.StatusCode, string(body))
	}

	if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
		return err
	}
	return nil
}

func ytDlpAvailable() bool {
	_, err := exec.LookPath("yt-dlp")
	return err == nil
}

func downloadAndProcessAudio(vodUrl string, startSec float64, outPcmPath string) error {
	clipStart := int(math.Max(0, startSec-5))
	clipEnd := clipStart + 10
	sectionArg := fmt.Sprintf("*%d-%d", clipStart, clipEnd)

	tmpDir, err := os.MkdirTemp("", "dl_*")
	if err != nil {
		return err
	}
	defer os.RemoveAll(tmpDir)

	dlBase := filepath.Join(tmpDir, "dl")

	cmd := exec.Command("yt-dlp",
		vodUrl,
		"--no-playlist",
		"--quiet",
		"--no-warnings",
		"--format", "bestaudio/best",
		"--download-sections", sectionArg,
		"-o", dlBase+".%(ext)s",
	)
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("yt-dlp failed: %v", err)
	}

	files, _ := filepath.Glob(dlBase + ".*")
	if len(files) == 0 {
		return fmt.Errorf("yt-dlp did not produce any audio file")
	}
	dlFile := files[0]

	cmdConv := exec.Command("ffmpeg",
		"-i", dlFile,
		"-ar", "2000",
		"-ac", "1",
		"-f", "s16le",
		"-c:a", "pcm_s16le",
		outPcmPath,
	)
	if err := cmdConv.Run(); err != nil {
		return fmt.Errorf("ffmpeg failed: %v", err)
	}

	return nil
}

func readPCM(path string) ([]int16, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	samples := make([]int16, len(b)/2)
	for i := 0; i < len(samples); i++ {
		samples[i] = int16(b[i*2]) | (int16(b[i*2+1]) << 8)
	}
	return samples, nil
}

func correlate(d1, d2 []int16) int {
	n1 := len(d1)
	n2 := len(d2)
	minLen := n1
	if n2 < minLen {
		minLen = n2
	}
	d1 = d1[:minLen]
	d2 = d2[:minLen]

	N := minLen
	bestLag := 0
	var maxCor int64 = math.MinInt64

	for lag := -N + 1; lag < N; lag++ {
		var sum int64
		if lag >= 0 {
			for i := 0; i < N-lag; i++ {
				sum += int64(d1[i+lag]) * int64(d2[i])
			}
		} else {
			absLag := -lag
			for i := 0; i < N-absLag; i++ {
				sum += int64(d1[i]) * int64(d2[i+absLag])
			}
		}

		if maxCor == math.MinInt64 || sum > maxCor {
			maxCor = sum
			bestLag = lag
		}
	}
	return bestLag
}

func HandleVodSync(w http.ResponseWriter, r *http.Request) {
	if r.Method != "POST" {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}
	w.Header().Set("Content-Type", "application/json")

	var req VodSyncRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "error": err.Error()})
		return
	}

	clientID := os.Getenv("TWITCH_CLIENT_ID")
	clientSecret := os.Getenv("TWITCH_CLIENT_SECRET")
	if clientID == "" || clientSecret == "" {
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "error": "Twitch API credentials not configured"})
		return
	}

	token, err := GetTwitchToken(clientID, clientSecret)
	if err != nil {
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "error": "Failed to authenticate with Twitch"})
		return
	}

	videoId, timeStr, totalSecs, clipSlug := parseTwitchUrl(req.Url)
	var absoluteTime time.Time

	sourceLabel := "Source VOD"
	cleanUrl := strings.ReplaceAll(req.Url, "https://", "")
	cleanUrl = strings.ReplaceAll(cleanUrl, "http://", "")
	cleanUrl = strings.ReplaceAll(cleanUrl, "www.", "")

	chMatch := regexp.MustCompile(`twitch\.tv/([^/?#]+)`).FindStringSubmatch(cleanUrl)
	if len(chMatch) > 1 {
		candidate := chMatch[1]
		if candidate != "videos" {
			sourceLabel = candidate
		}
	}

	// Determine the absolute timestamp of the source media
	if clipSlug != "" {
		var clipData TwitchClipResp
		err = doTwitchGet("https://api.twitch.tv/helix/clips?id="+clipSlug, clientID, token, &clipData)
		if err == nil && len(clipData.Data) > 0 {
			vId := clipData.Data[0].VideoId
			sourceLabel = clipData.Data[0].BroadcasterName

			if vId != "" {
				// Get video to find created_at
				var vData TwitchVideoResp
				err = doTwitchGet("https://api.twitch.tv/helix/videos?id="+vId, clientID, token, &vData)
				if err == nil && len(vData.Data) > 0 {
					createdAt, _ := time.Parse(time.RFC3339, vData.Data[0].CreatedAt)
					absoluteTime = createdAt.Add(time.Duration(clipData.Data[0].VodOffset) * time.Second)
				}
			} else {
				absoluteTime, _ = time.Parse(time.RFC3339, clipData.Data[0].CreatedAt)
			}
		}
	} else if videoId != "" {
		var vData TwitchVideoResp
		err = doTwitchGet("https://api.twitch.tv/helix/videos?id="+videoId, clientID, token, &vData)
		if err == nil && len(vData.Data) > 0 {
			sourceLabel = vData.Data[0].UserName
			createdAt, _ := time.Parse(time.RFC3339, vData.Data[0].CreatedAt)
			absoluteTime = createdAt.Add(time.Duration(totalSecs) * time.Second)
		}
	} else {
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "error": "Could not parse clip or video ID from URL"})
		return
	}

	var configs []VodSyncConfig
	configs = append(configs, VodSyncConfig{
		Label:       sourceLabel,
		Video:       videoId,
		Clip:        clipSlug,
		Time:        timeStr,
		Offset:      0,
		TotalOffset: 0,
	})

	// Loop over requested streamers and find synced VODs
	for _, streamer := range req.Streamers {
		// Get user ID
		var usersData struct {
			Data []struct {
				Id string `json:"id"`
			} `json:"data"`
		}
		endpoint := "https://api.twitch.tv/helix/users?login=" + url.QueryEscape(streamer)
		err = doTwitchGet(endpoint, clientID, token, &usersData)
		if err != nil || len(usersData.Data) == 0 {
			log.Printf("Could not find user %s", streamer)
			continue
		}
		userId := usersData.Data[0].Id

		// Get recent VODs
		var proxyData TwitchVideoResp
		endpoint = fmt.Sprintf("https://api.twitch.tv/helix/videos?user_id=%s&first=50", userId)
		err = doTwitchGet(endpoint, clientID, token, &proxyData)
		if err != nil {
			log.Printf("Could not fetch videos for %s", streamer)
			continue
		}

		var foundVod *TwitchVideoResp
		var foundVodIdx int
		for i, v := range proxyData.Data {
			vCreated, _ := time.Parse(time.RFC3339, v.CreatedAt)
			vDur := parseDurationStr(v.Duration)
			vEnded := vCreated.Add(time.Duration(vDur) * time.Second)

			// Video must have started before the target time and ended after it
			if (vCreated.Before(absoluteTime) || vCreated.Equal(absoluteTime)) && absoluteTime.Before(vEnded) {
				foundVodIdx = i
				foundVod = &proxyData // hacky but easy to reference
				break
			}
		}

		if foundVod != nil {
			v := foundVod.Data[foundVodIdx]
			vCreated, _ := time.Parse(time.RFC3339, v.CreatedAt)
			offsetSecs := absoluteTime.Sub(vCreated).Seconds()

			configs = append(configs, VodSyncConfig{
				Label:       streamer,
				Video:       v.Id,
				Time:        formatDurationStr(offsetSecs),
				Offset:      0.0,
				TotalOffset: 0.0,
			})
		}
	}

	if len(configs) >= 2 && ytDlpAvailable() {
		src := configs[0]
		if src.Video != "" {
			srcSecs := parseDurationStr(src.Time)
			srcVodUrl := "https://www.twitch.tv/videos/" + src.Video

			tmpDir, err := os.MkdirTemp("", "twitchsync_*")
			if err == nil {
				defer os.RemoveAll(tmpDir)

				srcPcmPath := filepath.Join(tmpDir, "src.pcm")
				err = downloadAndProcessAudio(srcVodUrl, srcSecs, srcPcmPath)
				if err == nil {
					srcAudio, _ := readPCM(srcPcmPath)
					for i := 1; i < len(configs); i++ {
						tgt := configs[i]
						if tgt.Video != "" {
							tgtSecs := parseDurationStr(tgt.Time)
							tgtVodUrl := "https://www.twitch.tv/videos/" + tgt.Video

							tgtPcmPath := filepath.Join(tmpDir, fmt.Sprintf("tgt_%d.pcm", i))
							err := downloadAndProcessAudio(tgtVodUrl, tgtSecs, tgtPcmPath)
							if err == nil {
								tgtAudio, _ := readPCM(tgtPcmPath)
								bestLag := correlate(srcAudio, tgtAudio)
								offsetSecs := float64(bestLag) / 2000.0
								correctedSec := tgtSecs - offsetSecs
								configs[i].Time = formatDurationStr(correctedSec)
								configs[i].Offset = offsetSecs
								configs[i].TotalOffset = correctedSec - srcSecs
							} else {
								log.Printf("Audio download failed for %s: %v", tgt.Label, err)
							}
						}
					}
				} else {
					log.Printf("Failed to download source audio: %v", err)
				}
			}
		}
	}

	json.NewEncoder(w).Encode(map[string]interface{}{
		"success": true,
		"data":    configs,
	})
}
